src/main/java/com/googleintern/wfm/ruleengine/action/RuleValidation.java

```diff
@@ -0,0 +1,286 @@
+    Files.deleteIfExists(outputFile.toPath());
+    outputFile.createNewFile();
+
+    FileWriter outputFileWriter = new FileWriter(outputFile);
+    CSVWriter csvWriter = new CSVWriter(outputFileWriter);
+
+    ImmutableList<String[]> validationResult = convertRuleValidationResultsToString();
+
+    csvWriter.writeAll(validationResult);
+    csvWriter.close();
+  }
+
+  public static double calculateRulesCoverage(
+      ImmutableSet<UserPoolAssignmentModel> usersWithWrongAssignedPermissions) {
+    double numberOfWrongAssignment = usersWithWrongAssignedPermissions.size();
+    double totalNumberOfUsers = expectedUserPoolAssignments.size();
+    return (totalNumberOfUsers - numberOfWrongAssignment) / totalNumberOfUsers;
+  }
+
+  public static ImmutableSet<UserPoolAssignmentModel> findUncoveredUsers() {
+    ImmutableSet.Builder<UserPoolAssignmentModel> usersWithWrongAssignedPermissionsBuilder =
+        ImmutableSet.builder();
+    for (UserPoolAssignmentModel user : expectedUserPoolAssignments) {
+      if (!actualAssignedPermissionsByUserPoolAssignment.containsKey(user)
+          && user.poolAssignments().size() != 0) {
+        usersWithWrongAssignedPermissionsBuilder.add(user);
+      } else if (!user.poolAssignments()
+          .equals(actualAssignedPermissionsByUserPoolAssignment.get(user))) {
+        usersWithWrongAssignedPermissionsBuilder.add(user);
+      }
+    }
+    return usersWithWrongAssignedPermissionsBuilder.build();
+  }
+
+  public static ImmutableSet<PoolAssignmentModel> findUncoveredPoolAssignments() {
+    ImmutableSet<PoolAssignmentModel> expectedAllPoolAssignments = findExpectedAllPoolAssignments();
+    ImmutableSet<PoolAssignmentModel> actualAllPoolAssignments = findActualAllPoolAssignments();
+    return ImmutableSet.copyOf(
+        Sets.difference(expectedAllPoolAssignments, actualAllPoolAssignments));
+  }
+
+  private static ImmutableSet<PoolAssignmentModel> findExpectedAllPoolAssignments() {
+    ImmutableSet.Builder<PoolAssignmentModel> expectedAllPoolAssignmentsBuilder =
+        ImmutableSet.builder();
+    for (UserPoolAssignmentModel user : expectedUserPoolAssignments) {
+      expectedAllPoolAssignmentsBuilder.addAll(user.poolAssignments());
+    }
+    return expectedAllPoolAssignmentsBuilder.build();
+  }
+
+  private static ImmutableSet<PoolAssignmentModel> findActualAllPoolAssignments() {
+    ImmutableSet.Builder<PoolAssignmentModel> actualAllPoolAssignmentsBuilder =
+        ImmutableSet.builder();
+    for (RuleModel rule : rules) {
+      for (Long permissionSetId : rule.permissionSetIds()) {
+        actualAllPoolAssignmentsBuilder.add(
+            PoolAssignmentModel.builder()
+                .setCasePoolId(rule.casePoolId())
+                .setPermissionSetId(permissionSetId)
+                .build());
+      }
+    }
+    return actualAllPoolAssignmentsBuilder.build();
+  }
+
+  private static ImmutableList<String[]> convertRuleValidationResultsToString() {
```

Comment:
toCSV? 

The return type you specified does not match the actual return type. 
Same below.